### PR TITLE
Add segment-wiki script

### PR DESCRIPTION
**Pull request by menshikh-iv**
Construct a corpus from a Wikipedia (or other MediaWiki-based) database dump and extract sections of pages from it.

CC @piskvorky

### CODE REVIEW

File: gensim/scripts/segment_wiki.py [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+    """
+    for article_title, article_sections in segment_all_articles(file_path):
+        printed_components = [json.dumps(article_title)]
+        for section_heading, section_content in article_sections:
+            printed_components.append(json.dumps(section_heading))
+            printed_components.append(json.dumps(section_content))
+        os.write(sys.stdout.fileno(), u"\t".join(printed_components).encode('utf-8') + b"\n")
+
+
+# noinspection PyUnresolvedReferences
```
Comment (piskvorky) [inline gensim/scripts/segment_wiki.py]: Noise, please remove (here and elsewhere).

**Comment by piskvorky** [comment]
@menshikh-iv the original script was Python3 only -- has this been tested on Python2?

We're aiming at dual compatibility (e.g. using `six`), like the rest of gensim.

**Comment by menshikh-iv** [comment]
@piskvorky of course, need to first check all another wiki scripts (as suggested in #1584), after it I'll add python2 compatibility here.

File: gensim/scripts/segment_wiki.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+            yield article_title, article_sections
+
+
+def segment_and_print_all_articles(file_path, output_file):
+    """
+    Prints article title and sections to stdout, tab-separated
+    article_title<tab>section_heading<tab>section_content<tab>section_heading<tab>section_content
+
+    """
+    with open(output_file, 'wb') as outfile:
```
Comment (piskvorky) [inline gensim/scripts/segment_wiki.py]: `smart_open`

File: gensim/scripts/segment_wiki.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+
+
+class WikiSectionsCorpus(WikiCorpus):
+    """
+    Treat a wikipedia articles dump (\*articles.xml.bz2) as a (read-only) corpus.
+
+    The documents are extracted on-the-fly, so that the whole (massive) dump
+    can stay compressed on disk.
+
+    >>> wiki = WikiCorpus('enwiki-20100622-pages-articles.xml.bz2') # create word->word_id mapping, takes almost 8h
```
Comment (piskvorky) [inline gensim/scripts/segment_wiki.py]: Docstring out of date (different class).

File: gensim/scripts/segment_wiki.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+            # file, so in practice we prune away enough.
+            elem.clear()
+
+
+def segment(page_xml):
+    """
+    Parse the content inside a page tag, returning its content as a list of tokens
+    (utf8-encoded strings).
+
+    Returns a 2-tuple (str, list) -
```
Comment (piskvorky) [inline gensim/scripts/segment_wiki.py]: Neither google nor numpy docstring format.

File: gensim/scripts/segment_wiki.py [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+    (title, [(section_heading, section_content)])
+
+    """
+    elem = cElementTree.fromstring(page_xml)
+    filter_namespaces = ('0',)
```
Comment (piskvorky) [inline gensim/scripts/segment_wiki.py]: Deserves a comment -- what is this?

**Comment by menshikh-iv** [comment]
About performance, approximately 5 minutes for 100,000 articles in the output file (SSD, i7 6700HQ, ruwiki), sometimes problems with `Ctrl+C` (if you want to interrupt), but it's non-critical.

File: gensim/scripts/segment_wiki.py (hunk 2 above)
Comment (menshikh-iv) [reply gensim/scripts/segment_wiki.py]: done

File: gensim/scripts/segment_wiki.py [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+        # is dumb and would load the entire input into RAM at once...
+        for group in utils.chunkize(page_xmls, chunksize=10 * self.processes, maxsize=1):
+            for article_title, sections in pool.imap(segment, group):  # chunksize=10):
+                # article redirects and short stubs are pruned here
+                num_total_tokens = 0
+                for section_title, section_content in sections:
+                    if self.lemmatize:
+                        num_total_tokens += len(utils.lemmatize(section_content))
+                    else:
+                        num_total_tokens += len(tokenize(section_content))
```
Comment (piskvorky) [inline gensim/scripts/segment_wiki.py]: Btw I think for the purposes of `gensim-data`, we shouldn't do any tokenization or normalization. We should present the sections "as they are", so people can use their own sentence detection / token detection etc. @menshikh-iv

Comment (menshikh-iv) [reply gensim/scripts/segment_wiki.py]: This is only for filtering very short articles, all content provided "as is".

File: gensim/scripts/segment_wiki.py (hunk 4 above)
Comment (menshikh-iv) [reply gensim/scripts/segment_wiki.py]: Done

File: gensim/scripts/segment_wiki.py (hunk 3 above)
Comment (menshikh-iv) [reply gensim/scripts/segment_wiki.py]: Done (updated all docstrings, converted to numpy-style, removed outdated things).

File: gensim/scripts/segment_wiki.py [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+        Parameters
+        ----------
+        fileobj : file
+            File descriptor of MediaWiki dump.
+        processes : int
+            Number of processes, max(1, multiprocessing.cpu_count() - 1) if None.
+        lemmatize : bool
+            If `pattern` package is installed, use fancier shallow parsing to get token lemmas.
+            Otherwise, use simple regexp tokenization.
+        filter_namespaces : tuple(int)
```
Comment (anotherbugmaster) [inline gensim/scripts/segment_wiki.py]: `tuple of int`

File: gensim/scripts/segment_wiki.py [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+
+    Parameters
+    ----------
+    file_path : str
+        Path to mediawiki dump, typical filename is <LANG>wiki-<YYYYMMDD>-pages-articles.xml.bz2
+        or <LANG>wiki-latest-pages-articles.xml.bz2.
+
+    Yields
+    ------
+    tuple(str, list of tuple(str, str))
```
Comment (anotherbugmaster) [inline gensim/scripts/segment_wiki.py]: `(str, list of (str, str))`

File: gensim/scripts/segment_wiki.py [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+    """Parse the content inside a page tag
+
+    Parameters
+    ----------
+    page_xml : str
+        Content from page tag.
+
+    Returns
+    -------
+    tuple(str, list of tuple(str, str))
```
Comment (anotherbugmaster) [inline gensim/scripts/segment_wiki.py]: `(str, list of (str, str))`

File: gensim/scripts/segment_wiki.py [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+
+        Note that this iterates over the **texts**; if you want vectors, just use
+        the standard corpus interface instead of this function::
+
+        >>> for vec in wiki_corpus:
+        >>>     print(vec)
+
+        Yields
+        ------
+        tuple(str, list of tuple(str, str))
```
Comment (anotherbugmaster) [inline gensim/scripts/segment_wiki.py]: `(str, list of (str, str))`

File: gensim/scripts/segment_wiki.py (hunk 6 above)
Comment (piskvorky) [reply gensim/scripts/segment_wiki.py]: Ah, OK, thanks.

**Comment by menshikh-iv** [comment]
Continued in #1694

**Pull request merged**